### PR TITLE
Port test cases in AbstractTransportTest to android interop test

### PR DESCRIPTION
**Pull request by zsurocking**
The test scenarios all remain the same. The changes are just in proto to nanoproto conversion and removing mocktio

### CODE REVIEW

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 1]
```diff
@@ -154,6 +193,21 @@ public void runTest(String testCase) throws Exception {
...
+      serverStreamingShouldBeFlowControlled();
+      veryLargeRequest();
+      veryLargeResponse();
+      deadlineNotExceeded();
+      deadlineExceeded();
+      deadlineExceededServerStreaming();
+      unimplementedMethod();
+      timeoutOnSleepingServer();
+      // This has to be the last one, because it will shut down the channel.
+      gracefulShutdown();
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: Add if clause below for each test to support running single test.

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 2]
```diff
@@ -274,47 +328,381 @@ public void pingPong() throws Exception {
...
-            responses.add(magicTailResponse);
-          }
-        };
+    ResponseObserver responseObserver = new ResponseObserver();
     StreamObserver<Messages.StreamingOutputCallRequest> requestObserver
         = asyncStub.fullDuplexCall(responseObserver);
     for (int i = 0; i < requests.length; i++) {
       requestObserver.onNext(requests[i]);
-      Object response = responses.poll(5, TimeUnit.SECONDS);
+      Object response = responseObserver.responses.poll(5, TimeUnit.SECONDS);
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: s/5/OPERATION_TIMEOUT

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 3]
```diff
@@ -274,47 +328,381 @@ public void pingPong() throws Exception {
...
-      assertEquals(goldenResponses[i], (Messages.StreamingOutputCallResponse) response);
-      Assert.assertTrue("More than 1 responses received for ping pong test.", responses.isEmpty());
+      assertMessageEquals(goldenResponses[i], (Messages.StreamingOutputCallResponse) response);
+      assertTrue("More than 1 responses received for ping pong test.",
+          responseObserver.responses.isEmpty());
     }
     requestObserver.onCompleted();
-    Assert.assertEquals(magicTailResponse, responses.poll(5, TimeUnit.SECONDS));
+    assertEquals(responseObserver.magicTailResponse,
+        responseObserver.responses.poll(5, TimeUnit.SECONDS));
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: ditto

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 4]
```diff
@@ -274,47 +328,381 @@ public void pingPong() throws Exception {
...
+  }
+
+  public void emptyStream() throws Exception {
+    @SuppressWarnings("unchecked")
+    ResponseObserver responseObserver = new ResponseObserver();
+    StreamObserver<StreamingOutputCallRequest> requestObserver
+        = asyncStub.fullDuplexCall(responseObserver);
+    requestObserver.onCompleted();
+    assertEquals(responseObserver.magicTailResponse,
+        responseObserver.responses.poll(5, TimeUnit.SECONDS));
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: ditto

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 5]
```diff
@@ -274,47 +328,381 @@ public void pingPong() throws Exception {
...
+    final StreamingOutputCallResponse goldenResponse = new StreamingOutputCallResponse();
+    goldenResponse.payload = new Payload();
+    goldenResponse.payload.type = Messages.COMPRESSABLE;
+    goldenResponse.payload.body = new byte[31415];
+
+    ResponseObserver responseObserver = new ResponseObserver();
+    StreamObserver<StreamingOutputCallRequest> requestObserver
+        = asyncStub.fullDuplexCall(responseObserver);
+    requestObserver.onNext(request);
+    Object response = responseObserver.responses.poll(5, TimeUnit.SECONDS);
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: ditto

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 6]
```diff
@@ -274,47 +328,381 @@ public void pingPong() throws Exception {
...
+    if (!(response instanceof Messages.StreamingOutputCallResponse)) {
+      fail("Unexpected: " + response);
+    }
+    assertMessageEquals(goldenResponse, (Messages.StreamingOutputCallResponse) response);
+
+    requestObserver.onError(new RuntimeException());
+    response = responseObserver.responses.poll(5, TimeUnit.SECONDS);
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: ditto

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 7]
```diff
@@ -274,47 +328,381 @@ public void pingPong() throws Exception {
...
+  }
+
+  public void cancelAfterBegin() throws Exception {
+    StreamRecorder<StreamingInputCallResponse> responseObserver = StreamRecorder.create();
+    StreamObserver<StreamingInputCallRequest> requestObserver =
+        asyncStub.streamingInputCall(responseObserver);
+    requestObserver.onError(new RuntimeException());
+    responseObserver.awaitCompletion();
+    assertEquals(Arrays.<StreamingInputCallResponse>asList(), responseObserver.getValues());
+    assertEquals(io.grpc.Status.CANCELLED,
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: Compare code instead.

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 2 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: done

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 3 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: done

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 4 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: done

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 5 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: done

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 6 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: done

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 7 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: done

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 8]
```diff
@@ -274,47 +356,381 @@ public void pingPong() throws Exception {
...
+      fail("Unexpected: " + response);
+    }
+    assertMessageEquals(goldenResponse, (Messages.StreamingOutputCallResponse) response);
+
+    requestObserver.onError(new RuntimeException());
+    response = responseObserver.responses.poll(OPERATION_TIMEOUT, TimeUnit.SECONDS);
+    if (!(response instanceof Throwable)) {
+      fail("Unexpected: " + response);
+    }
+    assertEquals(io.grpc.Status.CANCELLED, io.grpc.Status.fromThrowable((Throwable) response));
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: Compare code.

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 9]
```diff
@@ -274,47 +356,381 @@ public void pingPong() throws Exception {
...
+        .withDeadlineAfter(10, TimeUnit.MILLISECONDS);
+    try {
+      StreamingOutputCallRequest request = new StreamingOutputCallRequest();
+      request.responseParameters = new ResponseParameters[1];
+      request.responseParameters[0] = new ResponseParameters();
+      request.responseParameters[0].intervalUs = 20000;
+      stub.streamingOutputCall(request).next();
+      fail("Expected deadline to be exceeded");
+    } catch (Throwable t) {
+      assertEquals(io.grpc.Status.DEADLINE_EXCEEDED, io.grpc.Status.fromThrowable(t));
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: ditto

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java [hunk 10]
```diff
@@ -274,47 +356,381 @@ public void pingPong() throws Exception {
...
+      StreamingOutputCallRequest request = new StreamingOutputCallRequest();
+      request.payload = new Messages.Payload();
+      request.payload.body = new byte[27182];
+      requestObserver.onNext(request);
+    } catch (IllegalStateException expected) {
+      // This can happen if the stream has already been terminated due to deadline exceeded.
+    }
+
+    recorder.awaitCompletion();
+    assertEquals(io.grpc.Status.DEADLINE_EXCEEDED,
```
Comment (madongfly) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: ditto

File: android-interop-testing/app/src/main/proto/io/grpc/android/integrationtest/messages.proto [hunk 11]
```diff
@@ -136,3 +136,11 @@ message StreamingOutputCallResponse {
   // Payload to increase response size.
   Payload payload = 1;
 }
+
+// For reconnect interop test only.
+// Server tells client whether its reconnects are following the spec and the
+// reconnect backoffs it saw.
+message ReconnectInfo {
```
Comment (madongfly) [inline android-interop-testing/app/src/main/proto/io/grpc/android/integrationtest/messages.proto]: Seems this is not used in this PR?

File: android-interop-testing/app/src/main/proto/io/grpc/android/integrationtest/test.proto [hunk 12]
```diff
@@ -71,3 +71,16 @@ service TestService {
...
+
+// A simple service NOT implemented at servers so clients can test for
+// that case.
+service UnimplementedService {
+  // A call that no server should implement
+  rpc UnimplementedCall(grpc.testing.Empty) returns(grpc.testing.Empty);
+}
+
+// A service used to control reconnect server.
+service ReconnectService {
```
Comment (madongfly) [inline android-interop-testing/app/src/main/proto/io/grpc/android/integrationtest/test.proto]: Not used?

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 8 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: Done

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 9 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: Done

File: android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java (hunk 10 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java]: Done

File: android-interop-testing/app/src/main/proto/io/grpc/android/integrationtest/messages.proto (hunk 11 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/proto/io/grpc/android/integrationtest/messages.proto]: Removed

File: android-interop-testing/app/src/main/proto/io/grpc/android/integrationtest/test.proto (hunk 12 above)
Comment (zsurocking) [inline android-interop-testing/app/src/main/proto/io/grpc/android/integrationtest/test.proto]: Removed

**Comment by madongfly** [comment]
@zsurocking LGTM

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @zsurocking isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Pull request merged**